doc/sonic-build-system/SONiC-OpenSSL-FIPS-140-3.md

```diff
@@ -0,0 +1,140 @@
+
+## Table of Contents
+- [Abbreviation](#abbreviation)
+- [Requirement](#requirement)
+- [The cryptographic modules in SONiC](#the-cryptographic-modules-in-SONiC)
+- [SONiC FIPS Configuration](#SONiC-FIPS-Configuration)
+  * [Enable FIPS on system level](#Enable-FIPS-on-system-level)
+  * [Enable FIPS on application level](#Enable-FIPS-on-application-level)
+- [OpenSSL FIPS 140-3](#OpenSSL-FIPS-140-3)
+  * [OpenSSL Engine](#OpenSSL-Engine)
+  * [SymCrypt OpenSSL Engine](#symCrypt-openSSL-engine)
+  * [OpenSSL configuration for SymCrypt Engine](#OpenSSL-configuration-for-SymCrypt-Engine)
+  * [OpenSSL configuration enhancement](#OpenSSL-configuration-enhancement)
+  * [SymCrypt OpenSSL Engine debian package](#SymCrypt-OpenSSL-Engine-debian-package)
+- [Kerberos Cryptographic Module](#Kerberos-Cryptographic-Module)
+- [Golang Cryptographic Module](#Golang-Cryptographic-Module)
+
+
+## Abbreviation
+
+| Abbreviation | Description                                  |
+| ------------ | -------------------------------------------- |
+| CAVP         | Cryptographic Algorithm Validation Program   |
+| CST          | Cryptographic and Security Test              |
+| CMVP         | Cryptographic Module Validation Program      |
+| FIPS         | Federal Information Processing Standard      |
+
+## Requirement
+SONiC only uses cryptographic modules validated by FIPS 140-3, Make SONiC compliant with FIPS 140-3.
+
+## The cryptographic modules in SONiC
+
+| Module   | Use Scenarios | Description                                                |
+| -------------------- | --------------- | -------------------------------------------- |
+| OpenSSL              | Python, OpenSSH | Cyptography and SSL/TLS ToolKit              |
+| Kerberos             | OpenSSH         | Kerboros contains builtin crypto module      |
+| Golang               | sonic-restapi   | Golang contains builtin crypt module         |
+| Libgcrypto           | GPG             | A general purpose cryptographic library originally based on code from GnuPG |
+| Kernel Crypto        | --              | Linux crypto kernel module |
+
+In Scopes:
+1. OpenSSL
+2. Kerberos
+3. Golang
+
+Out of Scopes:
+1. Linux Kernel
+2. Libgcrypt
+
+## SONiC FIPS Configuration
+### Enable FIPS on system level
+Add the Linux System parameter fips=1, in grub config, one of implemetation as below:
+
+cat /etc/grub.d/99-fips.cfg
+```
+GRUB_CMDLINE_LINUX_DEFAULT="$GRUB_CMDLINE_LINUX_DEFAULT fips=1"
+```
+
+To validate the FIPS enabled, grep 'fips=1' /proc/cmdline.
+
+### Enable FIPS on application level
+```
+export ENABLE_FIPS=1
+```
+
+Alternative option for the golang applications only:
+```
+export GOLANG_FIPS=1
+```
+
+Alternative option for the OpenSSL applications only:
+
+see https://www.openssl.org/docs/manmaster/man7/openssl-env.html
+```
+export OPENSSL_CONFIG=/usr/lib/ssl/openssl-fips.cnf
+```
+
+## OpenSSL FIPS 140-3
+
+![FIPS Overview!](images/fips-overview.png)
```

Comment:
>png [](http://example.com/codeflow?start=39&length=3)

The "..." points to "Symcrypt Engine FIPS Module". Difficult to understand.

Reply:
In your diagram.

Reply:
The "..." means many applications using the cryptographic modules not listed out.
Only list some of the applications that will use them.